kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h

```diff
@@ -0,0 +1,1184 @@
+
+            // Fill transpose positions
+            for (IndexType i = 0; i < number_of_blocks; ++i) {
+                for (IndexType j = 0; j < number_of_blocks; ++j) {
+                    transpose_blocks(i, j) = false;
+                }
+            }
+
+            // Assemble the blocks
+            if (BaseType::mOptions.Is(DOUBLE_LAGRANGE_MULTIPLIER)) {
+                // Definition of the build scale factor auxiliar value
+                const bool has_auxiliar_constraint_scale_factor = mAuxiliarConstraintFactorConsidered == AUXILIAR_CONSTRAINT_FACTOR::CONSIDER_PRESCRIBED_CONSTRAINT_FACTOR ? true : false;
+                KRATOS_ERROR_IF(has_auxiliar_constraint_scale_factor && !r_current_process_info.Has(AUXILIAR_CONSTRAINT_SCALE_FACTOR)) << "Auxiliar constraint scale factor not defined at process info" << std::endl;
+                const double auxiliar_constraint_scale_factor = has_auxiliar_constraint_scale_factor ? r_current_process_info.GetValue(AUXILIAR_CONSTRAINT_SCALE_FACTOR) : mAuxiliarConstraintFactorConsidered == AUXILIAR_CONSTRAINT_FACTOR::CONSIDER_NORM_DIAGONAL_CONSTRAINT_FACTOR ? this->GetDiagonalNorm(rA) : this->GetAveragevalueDiagonal(rA);
+                mAuxiliarConstraintFactor = auxiliar_constraint_scale_factor;
+
+                // Create auxiliar identity matrix
+                TSystemMatrixType identity_matrix(number_of_slave_dofs, number_of_slave_dofs);
+                for (IndexType i = 0; i < number_of_slave_dofs; ++i) {
+                    identity_matrix.push_back(i, i, 1.0);
+                }
+
+                KRATOS_ERROR_IF_NOT(identity_matrix.nnz() == number_of_slave_dofs) << "Inconsistent number of non-zero values in the identity matrix: " << number_of_slave_dofs << " vs " << identity_matrix.nnz() << std::endl;
+
+                // Fill blocks
+                matrices_p_blocks(0,2) = &transpose_of_T;
+                matrices_p_blocks(2,0) = &copy_of_T;
+                matrices_p_blocks(1,1) = &identity_matrix;
+                matrices_p_blocks(1,2) = &identity_matrix;
+                matrices_p_blocks(2,1) = &identity_matrix;
+                matrices_p_blocks(2,2) = &identity_matrix;
+
+                // Fill coefficients
+                contribution_coefficients(0, 2) = mConstraintFactor;
+                contribution_coefficients(2, 0) = mConstraintFactor;
+                contribution_coefficients(1, 1) = -mAuxiliarConstraintFactor;
+                contribution_coefficients(1, 2) = mAuxiliarConstraintFactor;
+                contribution_coefficients(2, 1) = mAuxiliarConstraintFactor;
+                contribution_coefficients(2, 2) = -mAuxiliarConstraintFactor;
+
+                // Assemble the matrix (NOTE: Like the identity matrix is created inside the condition must be used meanwhile is alive, so inside the condition)
+                SparseMatrixMultiplicationUtility::AssembleSparseMatrixByBlocks(rA, matrices_p_blocks, contribution_coefficients, transpose_blocks);
```

Comment:
the `MultiplicationUtility` is doing assembly ? why don't you put it into the `space` ? It is misleading. By the wax does this work in MPI ?

Reply:
Yes, is doing the assembly. No, it only works in OpenMP

Reply:
It is not part of the space because includes many non standard things...

Reply:
ok .. then the naming and/or including an assembly method in MultiplicationUtility is not consistent.

Reply:
This method is already merged with the master, do you want me to modify the name that already exists and break its use in other implementations?

Reply:
I saw it used only in tests .. so I asked. If you say it will break other impl. Its fine ... !

Reply:
Only if this is supported by the @KratosMultiphysics/technical-committee , what name do you propose? The method basically is assembling a new sparse matrix from blocks